### PR TITLE
mgr/cephadm: fix the loki address in grafana, promtail configuration file

### PR DESCRIPTION
**Pull request by hookak**
- Fix to use loki address instead of MGR address(grafana datasource, promtail loki host)
- Add a loki dependency on grafana and promtail services

Signed-off-by: jinhong.kim <jinhong.kim0@navercorp.com>

reference: ba6efa121ab0e7b6949cf41219b68c38d1a77641

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 1]
```diff
@@ -445,14 +445,15 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
...
         deps: List[str] = []
-        hostnames: List[str] = []
-        for dd in self.mgr.cache.get_daemons_by_service('mgr'):
-            assert dd.hostname is not None
-            addr = self.mgr.inventory.get_addr(dd.hostname)
-            hostnames.append(addr)
+        daemons = self.mgr.cache.get_daemons_by_service('loki')
+        loki_host = ''
+        assert daemons is not None
+        if daemons != []:
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/monitoring.py:451]:
```suggestion
        if not daemons:
```
better way to check for empty list

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 2]
```diff
@@ -445,14 +445,15 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
...
+            assert daemons[0].hostname is not None
+            loki_host = daemons[0].ip if daemons[0].ip else self._inventory_get_addr(daemons[0].hostname)
```
Comment (rkachach) [inline src/pybind/mgr/cephadm/services/monitoring.py:453]: I think this is equivalent to:

```suggestion
            loki_host = daemons[0].ip or self._inventory_get_addr(daemons[0].hostname)
```

File: src/pybind/mgr/cephadm/services/monitoring.py (hunk 1 above)
Comment (hookak) [reply src/pybind/mgr/cephadm/services/monitoring.py:451]:
```suggestion
        if daemons:
```
for empty list check, it should be like this, isn't it?

File: src/pybind/mgr/cephadm/services/monitoring.py (hunk 2 above)
Comment (hookak) [reply src/pybind/mgr/cephadm/services/monitoring.py:453]: Thanks for your comment!

File: src/pybind/mgr/cephadm/services/monitoring.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/monitoring.py:451]: ah, yeah. My mistake. `if daemons` is correct for checking for nonempty list

Comment (hookak) [reply src/pybind/mgr/cephadm/services/monitoring.py]: appreciate for your review

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -2325,8 +2325,9 @@ def _calc_daemon_deps(self,
         else:
             need = {
                 'prometheus': ['mgr', 'alertmanager', 'node-exporter', 'ingress'],
-                'grafana': ['prometheus'],
+                'grafana': ['prometheus', 'loki'],
```
[the remark this answers is not in the document]
Comment (hookak) [reply src/pybind/mgr/cephadm/module.py:2328]: Is loki dependency likely to cause problems? In my opinion, this just reconfig the grafana service when the loki service is deployed.
Even if loki is not deployed, grafana works fine.
Please let me know if I'm missing

[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/cephadm/module.py:2328]: Ok, if that's the behavior, then makes sense. When Loki+Promtail services were added I wondered if we could state service dependencies or even create service bundles.

**Review by epuertat — COMMENTED** [review]
Thanks a lot @hookak for this fix! Are you using or planning to use Loki-Promtail to collect Ceph logs? Can you please tell us more about it? When we introduced this functionality we weren't fully sure about user adoption, so any feedback would be helpful. Thanks!

@aaSharma14 can you please review & test this? I don't think we had any automated tests for loki-promtail deployment, right?

File: src/pybind/mgr/cephadm/module.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:2328]: yeah, to clarify, all this "dependency" does is reconfig the daemon if we place or remove a daemon whose type it is "dependent" on.

**Comment by hookak** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 4]
```diff
@@ -445,14 +444,14 @@ def prepare_create(self, daemon_spec: CephadmDaemonDeploySpec) -> CephadmDaemonD
...
-        hostnames: List[str] = []
-        for dd in self.mgr.cache.get_daemons_by_service('mgr'):
-            assert dd.hostname is not None
-            addr = self.mgr.inventory.get_addr(dd.hostname)
-            hostnames.append(addr)
+        daemons = self.mgr.cache.get_daemons_by_service('loki')
+        loki_host = ''
+        if daemons:
+            assert daemons[0].hostname is not None
+            loki_host = daemons[0].ip or self._inventory_get_addr(daemons[0].hostname)
```
Comment (aaSharma14) [inline src/pybind/mgr/cephadm/services/monitoring.py:451]: On testing this PR, Got this error - `Failed to apply: 'PromtailService' object has no attribute '_inventory_get_addr'` on starting the Promtail Service

**Comment by adk3798** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/services/monitoring.py (hunk 4 above)
Comment (hookak) [reply src/pybind/mgr/cephadm/services/monitoring.py]: Thanks for comment!!

Comment (aaSharma14) [reply src/pybind/mgr/cephadm/services/monitoring.py]: Works fine now. Thank @hookak

**Comment by adk3798** [comment]
@hookak there is a unit test failing. I'm guessing it's some automation from your text editor removing ending whitespace or something but it's causing the test to fail. Putting the space back fixed it

image

**Comment by hookak** [comment]
> @hookak there is a unit test failing. I'm guessing it's some automation from your text editor removing ending whitespace or something but it's causing the test to fail. Putting the space back fixed it
> 
> image

Appreciated! This pr failed to 'ceph dashboard cephadm e2e test', is it related to?

**Comment by adk3798** [comment]
> > @hookak there is a unit test failing. I'm guessing it's some automation from your text editor removing ending whitespace or something but it's causing the test to fail. Putting the space back fixed it
> > image
> 
> Appreciated! This pr failed to 'ceph dashboard cephadm e2e test', is it related to?

@hookak I just checked the output of the dashboard cpehadm test and saw
image

I don't think it's related to your changes. I believe that test is just not stable currently.

**Comment by adk3798** [comment]
jenkins test dashboard cephadm

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-07-17_04:25:04-orch:cephadm-wip-adk-testing-2022-07-16-1847-distro-default-smithi/

2 failures:

- 1 tracked by https://tracker.ceph.com/issues/56552
- the agent/on test_nfs test which has been failing in various ways for a while now

Neither failure is new or related to PRs in the run

**Comment by adk3798** [comment]
@hookak would you like this to be backported to quincy?

**Comment by hookak** [comment]
Yes, I would. Thanks a lot! @adk3798

**Pull request merged**